src/quickgui/qgsquickmapsettings.cpp

```diff
@@ -0,0 +1,224 @@
+/***************************************************************************
+  qgsquickmapsettings.cpp
+  --------------------------------------
+  Date                 : 27.12.2014
+  Copyright            : (C) 2014 by Matthias Kuhn
+  Email                : matthias (at) opengis.ch
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+
+#include "qgsmaplayer.h"
+#include "qgsmaplayerstylemanager.h"
+#include "qgsmessagelog.h"
+#include "qgsproject.h"
+#include "qgis.h"
+
+#include "qgsquickmapsettings.h"
+
+QgsQuickMapSettings::QgsQuickMapSettings( QObject *parent )
+  : QObject( parent )
+{
+  // Connect signals for derived values
+  connect( this, &QgsQuickMapSettings::destinationCrsChanged, this, &QgsQuickMapSettings::mapUnitsPerPixelChanged );
+  connect( this, &QgsQuickMapSettings::extentChanged, this, &QgsQuickMapSettings::mapUnitsPerPixelChanged );
+  connect( this, &QgsQuickMapSettings::outputSizeChanged, this, &QgsQuickMapSettings::mapUnitsPerPixelChanged );
+  connect( this, &QgsQuickMapSettings::extentChanged, this, &QgsQuickMapSettings::visibleExtentChanged );
+  connect( this, &QgsQuickMapSettings::rotationChanged, this, &QgsQuickMapSettings::visibleExtentChanged );
+  connect( this, &QgsQuickMapSettings::outputSizeChanged, this, &QgsQuickMapSettings::visibleExtentChanged );
+}
+
+void QgsQuickMapSettings::setProject( QgsProject *project )
+{
+  if ( project == mProject )
+    return;
+
+  // If we have already something connected, disconnect it!
+  if ( mProject )
+  {
+    disconnect( mProject, nullptr, this, nullptr );
+  }
+
+  mProject = project;
+
+  // Connect all signals
+  if ( mProject )
+  {
+    connect( mProject, &QgsProject::readProject, this, &QgsQuickMapSettings::onReadProject );
+
+    // TODO: we have a problem here, since project can have alread .qgs loaded, so
+    // we are unable to get DOM to reload project settings for the canvas
+    // fortunately setProject is used only once at the very beginning of the run,
+    // so it should work ....
```

Comment:
Was that me who wrote this comment? I'm having a hard time understanding what I wanted to tell the world :D

Reply:
it was someone over here, probably me. The problem arose with removal of `QgsProject::instance`.

Bug is when you  do something like 
```
QgsProject p;
p.read()
qgsquickmapsettings.setProject(p);
```
and in this case QgsQuickMapSettings::onReadProject is not triggered, so it is not possible to read mapcanvas settings from DOM

Reply:
And if one would just trigger an onReadProject() here?

Reply:
you do not have DOM string any more.

Reply:
It's been a while since I wrote all this... Why are things read directly from the DOM? Are they not available elsewhere?

Reply:
But the parameter is QgsProject, and you do not have any other mapsettings instance but internal one in the qgsquickmapsettings... not sure from where we could read it?

Reply:
That's what I'm asking you.
If it's not possible (could be) I think the TODO comment can be removed (and possibly replaced with a comment for pre-conditions).

Reply:
It is not possible without some refactoring of QgsProject. Moved comment to header for now to state prerequisite.